Reorganize import/export docs (#18761)
mdn/content: files/en-us/web/javascript/reference/statements/export/index.md

---
{{jsSidebar("Statements")}}

> The **`export`** declaration is used when creating JavaScript modules to export values from the module so they can be used by other programs with the {{jsxref("Statements/import", "import")}} declaration or [dynamic import](/en-US/docs/Web/JavaScript/Reference/Operators/import). The value of an imported binding is subject to change in the module that exports it. When a module updates the value of a binding that it exports, the update will be visible in its imported value.

Reviewer: Consider shorter sentences.

```suggestion
The **`export`** declaration is used to export values from a JavaScript module.
Exported values can then be imported into other programs with the {{jsxref("Statements/import", "import")}} declaration or [dynamic import](/en-US/docs/Web/JavaScript/Reference/Operators/import).
The value of an imported binding is subject to change in the module that exports it.
When a module updates the value of a binding that it exports, the update will be visible in its imported value.
```

In order to use the `export` declaration in a source file, the file must be interpreted by the runtime as a [module](/en-US/docs/Web/JavaScript/Guide/Modules). In HTML, this is done by adding `type="module"` to the 

> {{HTMLElement("script")}} tag, or by being imported by another module. Modules are automatically interpreted in [strict mode](/en-US/docs/Web/JavaScript/Reference/Strict_mode).

Reviewer: "Modules are automatically interpreted in [strict mode](/en-US/docs/Web/JavaScript/Reference/Strict_mode)."

This is better than what it replaces but I still don't know what it means. Are you saying that you don't need to specify module exports in order to export /import a module? IN this case, what then is exported? (link to more info maybe).

Reviewer: OK. I was overthinking this.

## Syntax

```js
// Exporting declarations
export let name1, name2/*, … */; // also var
export const name1 = 1, name2 = 2/*, … */; // also var, let

> export function functionName() { /* … */ } // also class, function*

Reviewer: "also class, function* "
-  why the asterisk at the end?
- if this is exporting a function, why mention function?
- I would just split out class into its own line too.

Reviewer: I need to get out more.

Personally I would split this out into three. Why? Because I missed that comment. And while I would assume that you could export var, let and const from one example (as in the previous line) I would not assume that being able to export a function implies you can export a class or a generator function (and one day I will find out what that is :-) 

Up to you.

## Syntax

There are two types of exports:

1. Named Exports (Zero or more exports per module)
2. Default Exports (One per module)

```js
// Exporting individual features
export let name1, name2, …, nameN; // also var, const
export let name1 = …, name2 = …, …, nameN; // also var, const
export function functionName(){...}
export class ClassName {...}
// Exporting declarations
export let name1, name2/*, … */; // also var
export const name1 = 1, name2 = 2/*, … */; // also var, let
export function functionName() { /* … */ } // also class, function*
export const { name1, name2: bar } = o;
export const [ name1, name2 ] = array;

// Export list
export { name1, name2, …, nameN };

> // Renaming exports

Reviewer: Should these be under the Export list heading? Some of them do feel like renaming, and that is a useful distinction.

Author: Renaming is part of the features of "export list".

> export { name1 as default /*, … */ };

Reviewer: I haven't read far enough to understand all this yet, but perhaps this should go under the next heading "Default exports"

Author: Yeah, this one is indeed "default export", but it's "default export using a named-export-like syntax" 😄 I think putting it in either place is fine—"default export" highlights its semantics while "named export" highlights its syntax

Reviewer: Cool, leave it where it is or duplicate in both places. Up to you :-)

## Description

Every module can have two different types of export, _named export_ and _default export_. You can have multiple named exports per module but only one default export. Each type corresponds to one of the above syntax.

> Named exports:

Reviewer: IN a line a few below it says `// export features declared earlier`. When is earlier.

Don't have to answer this here - but we should make a statement about where export might appear.

Author: "Earlier" is not the right word. Should be "elsewhere".

Reviewer: Cool. Note "but we should make a statement about where export might appear." - is that clearly stated?

> export function myFunction() { /* ... */ };

Reviewer: Is there a mismatch indentation with above line? Might just be the review pane

Author: Yes—because of the diff markers, changed lines will be pushed in by two characters' width.

```

> After the `export` keyword, you can use `let`, `const`, and `var` declarations, as well as function or class declarations. You can also use the `export { name1, name2 }` syntax to export a list of names declared elsewhere. Note that `export {}` does not export an empty object — it's a no-op declaration that exports nothing (an empty name list).

Reviewer: Is there anything special we need to think about when exporting let, const? I.e. in terms of the scoping and temporal dead zone etc of "normal use"?

Author: Good question! I'll add that 👍

Default exports:

```js
// export feature declared earlier as default
export { myFunction as default };
// This is equivalent to:
export default myFunction;

// export individual features as default
export default function () { /* … */ }
export default class { /* … */ }

// each export overwrites the previous one
```

> The `export default` syntax allows any expression. As a special case, functions and classes are exported as _declarations_ that can be anonymous, which means functions will be hoisted.

Reviewer: I don't understand the special case or what it means to say "that can be anonymous"

 I.e. a function that isn't a function expression is always hoisted right?
 
Or to put it another way, below there is an example that works, but I don't understand why anyone would think it wouldn't work. Does that make sense?

Author: There are two things: function expressions and function declarations.

```js
// declaration
function foo() {}

// expression
const a = function foo() {}
```

The `export default` accepts expressions, so at first sight, this looks like it should be exporting an expression:

```js
export default function foo() {}
```

But nope, this is actually a _declaration_, both by spec terms and by how it actually works—because only declarations are hoisted.

But it's a special kind of declaration—namely, it's the only place where you can omit the name! If you write:

```js
function () {}
```

It's a syntax error because declarations can't be unnamed (since there's no way to refer to them anyway). But exported declarations can be unnamed. This is what "declarations that can be anonymous" means. It's a pretty special case.

Reviewer: This is an excellent explanation thank you. 

I do understand the difference between function expressions and declarations. What I hadn't understood was that someone looking at `export default function foo() {}` might think that you're exporting a variable with foo assigned - i.e. a function expression, but this is actually a declaration. 

It might be worth expanding on this. In particular the anonymous thing. Presumably you can export it anonymous because it is default - which has an implicit name on import.

Up to you, but I even though I understand this now, I still think someone else reading this will have the same question.

```js
// Works because `foo` is a function declaration,
// not a function expression
foo();

export default function foo() {
  console.log("Hi");
}
```

> Named exports are useful to export several values. During the import, it is mandatory to import them within curly braces with the same name of the corresponding object. But a default export can be imported with any name. For example:

Reviewer: I didn't understand the second sentence. Is this what you mean?

```suggestion
Named exports are useful when you need to export several values.
Named exports must be imported within curly braces and using the **same name** as was used in the export expression.
A default export can be imported with any name.
For example:
```

};
```

> You can rename a name to something that's not a valid identifier by using a string literal. For example:

Reviewer: That's interesting. When might I want to do this? (do we need to say?)

Author: I don't think it's particularly useful within the ES module scope, but ES modules are designed with interoperation in mind, and existing CommonJS modules (which Node.js uses) more commonly utilize non-identifier names. For example:

```js
module.exports.function = function () {};
module.exports["something-interesting"] = function () {};
```

If we don't allow non-identifiers (and by identifier we probably include reserved words as well), then there's no way to convert these to ES modules.

Reviewer: OK. Then I don't think this needs more said. In any case, more is said further on.

```js
export { myFunction as "my-function" };
```

### Re-exporting / Aggregating

> It is also possible to "import/export" from different modules in a parent module so that they are available to import from that module. In other words, one can create a single module concentrating various exports from various modules.

Reviewer: The first sentence confuses me. Do you mean:

```suggestion
It is also possible to create a single module aggregating various exports from various modules.
This is done by exporting from a number of  "child" modules that are then imported and re-exported from a parent module.
```

Reviewer: Perhaps not. Reading further the second sentence might be

> This is done by a parent module _exporting from_ a number of  "child" modules.

The difference being that the import is kind of invisible. I'll leave this with you, but the paragraph definitely needs work

Author: This sentence is indeed too confusing. I'll reword it.
